src/bitcoin/Transaction.mo

```diff
@@ -133,49 +200,310 @@ module {
+          witnessesBuffer.add(Witness.toBytes(witnesses[i]));
+        };
+      };
+
+      let serializedWitnesses = Array.flatten(Buffer.toArray(witnessesBuffer));
+
+      // Total size of output excluding sigHashType.
+      let totalSize : Nat =
+      // 4 bytes for version.
+      4
+      // 2 additional bytes if witness is present.
+      + maybeAdditionalWitnessFlags.size()
+      // transaction inputs and outputs
+      + serializedTxInSize.size() + totalTxInSize + serializedTxOutSize.size() + totalTxOutSize
+      // serialized witnesses if any
+      + serializedWitnesses.size()
+      // 4 bytes for locktime.
+      + 4;
+      let output = Array.init<Nat8>(totalSize, 0);
+      var outputOffset = 0;
+
+      // Write version.
+      Common.writeLE32(output, outputOffset, version);
+      outputOffset += 4;
+
+      Common.copy(
+        output,
+        outputOffset,
+        maybeAdditionalWitnessFlags,
+        0,
+        maybeAdditionalWitnessFlags.size(),
+      );
+      outputOffset += maybeAdditionalWitnessFlags.size();
+
+      // Write TxInputs size.
+      Common.copy(
+        output,
+        outputOffset,
+        serializedTxInSize,
+        0,
+        serializedTxInSize.size(),
+      );
+      outputOffset += serializedTxInSize.size();
+
+      // Write serialized TxInputs.
+      for (serializedTxIn in serializedTxIns.vals()) {
+        Common.copy(
+          output,
+          outputOffset,
+          serializedTxIn,
+          0,
+          serializedTxIn.size(),
+        );
+        outputOffset += serializedTxIn.size();
+      };
+
+      // Write TxOutputs size.
+      Common.copy(
+        output,
+        outputOffset,
+        serializedTxOutSize,
+        0,
+        serializedTxOutSize.size(),
+      );
+      outputOffset += serializedTxOutSize.size();
+
+      // Write serialized TxOutputs.
+      for (serializedTxOut in serializedTxOuts.vals()) {
+        Common.copy(
+          output,
+          outputOffset,
+          serializedTxOut,
+          0,
+          serializedTxOut.size(),
+        );
+        outputOffset += serializedTxOut.size();
+      };
+
+      Common.copy(
+        output,
+        outputOffset,
+        serializedWitnesses,
+        0,
+        serializedWitnesses.size(),
+      );
+      outputOffset += serializedWitnesses.size();
+
+      // Write locktime.
+      Common.writeLE32(output, outputOffset, locktime);
+      outputOffset += 4;
+
+      assert (outputOffset == output.size());
+      let result = Array.freeze(output);
+      result;
+    };
+
+    // Serialize transaction to bytes with layout:
+    // | version | witness flags if it is present | len(txIns) | txIns |
+    // len(txOuts) | txOuts | witnesses | locktime |
+    //
+    // This function is required to compute the transaction id if it contains a
+    // witness, since the id is a hash over the serialized transaction ignoring the witness.
+    public func toBytesIgnoringWitness() : [Nat8] {
```

Comment:
If you have a concrete implementation in mind, then feel free to describe it, but I myself don't see how this can be nicely unified without making it even less readable.

Reply:
At a quick glance, it looked like one could (on a high level) let the `toBytes` impl take a flag `include_witness: bool` and then not count/serialize all witness-related parts. But, as I said, this is somewhat optional.